fix!: Major improvements to `kraft cloud compose`

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR includes major fixes and improvements which increase the compatibility of using `kraft cloud compose` against existing compose projects.

This series of commits largely include numerous fixes to the creation (building) of images as well as some partial refactors to the `kraft cloud compose up` subcommand for better compatibility.  Additionally, the following new features are available:

- Dynamically building working directories and projects without the need for a `Kraftfile` and only the presence of a `Dockerfile`;
- Introduce a `--runtime` flag allowing users to specify an alternative pre-built Unikraft unikernel runtime to be used when images are dynamically built from a `Dockerfile`;
- Introduce a `--no-build` flag to `kraft cloud compose up` to prevent the build step from occuring before instance instantiation;
- Preliminary support for an `--entrypoint` flag which is used to signal when to use arguments supplied in the existing OCI image and when to overwrite them.

nderjung committed:
fix(compose): Do not associate networks with service groups

This refactor removes the association of "compose networks" with
KraftCloud service groups.  This association is in fact incorrect
as the service groups are intended to be used for a single
application as opposed to a group of networks; to be used for
autoscaling and load-balancing for said single application.

As a result of this refactor, introduce a new non-attribute
`Services` to the compose create routine such that we can still
propagate ports correctly.

Additionally, adjust the `kraft cloud compose down` CLI subcommand
as it does not need to remove associated service groups anymore.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor: Remove unused `error` return

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor: Remove unused `error` return

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(app): Remove unused typecasting

Prefer the abstraction interface.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor: Remove unused `error` return

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(app): Use variadic arguments for `WithTargets`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(build): Expose `Project` attribute

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(build): Only instantiate `Project` if `nil`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(app): Convert `application` interface methods to pointer

Allows for non-immutable changes to the structure.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(app): Add `SetRootfs` to `application` object

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(runtime): Declare default runtime for KraftCloud

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(runtime)!: Rename constant

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented:
What would be the alternative to this then? Using the internal IP everywhere?

nderjung commented:
All instances in your account can communicate via internal networking.

nderjung committed:
refactor(kraft): Rework `BuildRootfs` to accept string arch arg

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Make the parent directory before saving

In case it does not exist.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Also save hard links

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Prevent recursion by only visiting non-links

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Remove rogue log warning

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(build): Contain project initialization within builder

In the future we aim to support dynamically instantiated `Project`
attributes for different in-memory usecases.  Prevent instantiation
by default by instead using the existing `initProject` when this is
necessary for contextualization.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(build): Set the rootfs for a project after build

Set the root file system for the project, since typically a
packaging step may occur after a build, and the root file system
is required for packaging and the packaging step may perform a
build of the rootfs again.  Ultimately this prevents re-builds.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Remove unused `copts` variable

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Only throw connection warning on connection errors

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Warn on failure to create hardlink

Instead of error-ing out, simply warn the hard link could not
be created.  It turns out, more often than not, the issue is
that the Dockerfile has a hardlink src which does not exist.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(initrd): Inject more debug logs

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Syntactically capture unused return error value

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Do not prepend contents of `initrd.args`

This variable is empty anyway.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Correct logic for setting architecture choice

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Ignore errors with cache exports

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Correct path to dockerfile in local directories

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Fix handling non-existing and absolute paths

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(target): Introduce default target for KraftCloud

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(app): Wrap commmand in string slice, do not split

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pkg): Do not pass args via shell escape

Since `opts.Args` can be passed from multiple sources including
the command-line, this unnecessarily escape quotes in most
usecases.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pkg): Introduce `Dockerfile` packager

This new packager constructs a project without a Kraftfile
simply from a `Dockerfile` detected as the rootfs.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(build): Introduce `Dockerfile` builder

This new builder constructs a project without a Kraftfile
simply from a `Dockerfile` detected as the rootfs.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
gomod(deps): Bump sdk.kraft.cloud from v0.5.9 to v0.5.10

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(cmdfactory): Recognize `types.ShellWords` in the command-line

This type is specifically used within compose and Docker's internal
tooling such that it can be recognised a). as its own type and b).
such that it is nil-able without it being empty.  Unlike a
`[]string` type, `nil` also signals that the slice is of size 0.
In this case, we want to also detect that it is not set at the
command-line.  This will allow commands to use the type and
know whether it is nil, it has been set and is specifically empty
or is non-empty.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(create): Introduce `--entrypoint` flag

This is used to overwrite the image's arguments.  Similar to
Docker and `Dockerfile`'s, the image has a supplied default command
which can be referenced and used.  With the use of
`types.ShellCommand` we can also detect whether an empty set of
command should be used.

The new default functionality is to accept the set of positional
arguments as appends to the images arguments.  The user can
overwrite this by setting a value to `--entrypoint`.  Supplying
an empty `--entrypoint` zeros-out that image value.

Fetch the metadata of the image beforehand to ensure we correctly
set everything too.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(create): Do not use (sub)domain if value is empty

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(create): Set arbitrary memory limit

Set the default memory to the size of the image if not specified
plus an arbitrary 10% buffer.  Only set the value if it is
greater than 128 MiB.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(create): Set the image on the right-hand side of the processtree

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(build): Correct usage struct tag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(build): Add `--push` usage example

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(build): Better handling of compose builds

This fix adds more logic to the `kraft cloud compose build`
subcommand to better stabilize it.  Specifically, it tries
to handle the case where the compose spec does not provide
a "build" which translates to it referencing an image.
However, these images may be valid OCI images which are not
unikernels.  A look up is performed to determine whether
there is a KraftCloud version (both from the user and
officially) of this image equivilant.  If neither, then
a build (or at least a "repackage") _can_ be performed
by this toolchain in order to convert the traditional
OCI image into a valid unikernel OCI.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(compose): Introduce `--runtime` switch flag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(logs): Quietly handle errors and check state more often

The log tailing can continue forever even after an instance
has stopped.  As a result, always check the state.  Whilst this
does introduce an additional call, it does allow the tailing
to quit earlier faster.

Additionally, provide the reason why the instance may have stopped.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(up): Pass the entrypoint from the compose file

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(up): Provide more thorough check for images

Additionally, update the service within the compose object
which is later passed to the `Create` method, such that it
has a correct image name reference.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(up): Introduce `--no-build` flag

This is also compliant with `docker compose up`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(up): Correct memory limit and reservation size representation

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(up): Prefix project name to service name

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(down): Use project + service name or container name if provided

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(logs): Use project + service name or container name if provided

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(stop): Use project + service name or container name if provided

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(ps): Use project + service name or container name if provided

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on initrd/ociimage.go:
```diff
@@ -213,7 +213,7 @@ func (initrd *ociimage) Build(ctx context.Context) (string, error) {
 		// whether the path is the same as the reference path which indicates
 		// whether the entry is a symbolic link.
-		if f.FileType == scfile.TypeSymLink {
+		if f.FileType == scfile.TypeSymLink || f.FileType == scfile.TypeHardLink {
 			header.Mode |= cpio.TypeSymlink
```
Are you 100% sure this is correct? You are marking the HardLink as a SymLink

nderjung replied:
Good point; i will double check.

nderjung replied:
This should hopefully now be resolved through the latest commit.

craciunoiuc commented on internal/cli/kraft/cloud/instance/create/create.go:
```diff
@@ -191,6 +192,13 @@ func Create(ctx context.Context, opts *CreateOptions, args ...string) (*kcclient
+	} else {
+		// Set the default memory to the size of the image if not specified plus an
+		// arbitrary 10% buffer.  Only set the value if it is greater than 128 MiB.
+		mem := int(math.Round(float64(image.SizeInBytes / 1024 / 921)))
```
This looks wrong, why are you dividing KiB by `921` and not `1024` to get MiB

I Get the 10% buffer, but you can just multiply by `1.1` after

(Go anyway transforms the divisions you a single op for efficiency) (I think)

craciunoiuc replied:
Moreover, remember we are still using bbuddy, which will waste 1/2 of the memory it allocates, so 10% is not enough for most cases

nderjung replied:
Will double check which allocator we're using. Funny though, this was enough for my testing to work.

I can change it to `(n/1024/1024)*1.1` to make it more easily readable.

nderjung replied:
But what i can do is just increase the memory size to, instead, to the nearest 2^N.

craciunoiuc replied:
Actually, just leave the 10% buffer, we can just pray to Simon to get the other allocator in

craciunoiuc commented on internal/cli/kraft/cloud/compose/up/up.go:
```diff
@@ -173,6 +177,45 @@ func Up(ctx context.Context, opts *UpOptions, args ...string) error {
+			officialPkgName = strings.ReplaceAll(service.Image, "_", "-")
+		} else {
+			userPkgName = fmt.Sprintf(
+				"%s/%s-%s:latest",
```
This is not prefixed with `index.unikraft.io`?

nderjung replied:
The API does not like this. Additionally, `index.unikraft.io` (+`/official`) is stripped in the next `create.Create` method anyway.

craciunoiuc replied:
okay

nderjung committed:
fix(build): Also check if the image exists before performing a build

If the build context has not been defined, we can still infer that an
image exists by performing a check and correctly associating it.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(up): Switch the check order: prefer official images over user images

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(initrd): Rework initrd creation and ensure integrity

This large commit reworks the creation of initrd CPIO archives
for both OCI images and Dockerfiles which simultaneously fixes
a number of known issues as well as ensures these issues no
longer persist with the introduction of relevant unit tests.

Issues which existed prior to this commit and are resolved
include:

- Unused `Files` interface method which was not always
  properly populated;
- Inability to correctly handle hardlinks both in
  Dockerfiles and OCI images;
- Inability for OCI images to properly reference
  symolic links.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
How bad can it go?

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
